ConcordiumWallet/Service/MobileWallet.swift

```diff
@@ -102,46 +174,15 @@ class MobileWallet: MobileWalletProtocol {
-                        self.storageManager.removeAccount(account: initialAccount)
-                        
-                        // cleanup if there have been other unsucessful tries and if the name was not set
-                        let failedAccounts = self.storageManager.getAccounts(for: identity)
-                        for account in failedAccounts {
-                            if initialAccountName == "" {
-                                initialAccountName = account.name ?? ""
-                            }
-                            self.storageManager.removeAccount(account: account)
-                        }
-                        
-                        // save account
-                        account.name = initialAccountName
-                        account.address = data.initialAccountData.accountAddress
-                        account.encryptedBalanceStatus = .decrypted
-                        account.encryptedAccountData = try self.storageManager.storePrivateAccountKeys(data.initialAccountData.accountKeys,
-                                                                                                       pwHash: pwHash).get()
-                        account.encryptedPrivateKey = try self.storageManager.storePrivateEncryptionKey(data.initialAccountData.encryptionSecretKey,
-                                                                                                        pwHash: pwHash).get()
-                        
-                        _ = try self.storageManager.storeAccount(account)
-                    } catch {
-                        return .fail(error)
-                    }
-                    return self.storageManager.storePrivateIdObjectData(data.privateIDObjectData.value, pwHash: pwHash).onSuccess {
-                        _ = identity.withUpdated(encryptedPrivateIdObjectData: $0)
-                        let newAccount = account.withUpdatedIdentity(identity: identity)
-                        let shieldedAmount = ShieldedAmountTypeFactory.create().withInitialValue(for: newAccount)
-                        _ = try? self.storageManager.storeShieldedAmount(amount: shieldedAmount)
-                    }.publisher.eraseToAnyPublisher()
-            }
-            .map { _ in
-                data.idObjectRequest
-            }.eraseToAnyPublisher()
+                .tryMap { (pwHash) in
+                    let identityCreation = try IdentityCreation(initialAccountName: initialAccountName,
+                                                            identityName: identityName,
+                                                            identityProvider: identityProvider,
+                                                            data: data,
+                                                            pwHash: pwHash,
+                                                            storageManager: self.storageManager)
```

Comment:
Since you capture the `self` weakly it becomes optional:

```suggestion
                                                            storageManager: self?.storageManager)
```